File: src_simple/f90/schmidt.f90
! schmidt  -  13.03.2014

! ******************************************
! Procédé d'orthogonalisation de Gram-Schimdt.
! Soit u et v deux vecteurs. On cherche le vecteur vp 
! le plus proche de v orthogonal à u.
! ******************************************

program schmidt

    implicit none

    integer            :: i
    real               :: scalaire, normu
    real, dimension(3) :: u, v, vp

    ! vecteur u
    u(1) = 1.0
    u(2) = 0.0
    u(3) = 0.0
    normu = sqrt(sum((u(:))**2))

    ! vecteur v
    v(1) = 1.0
    v(2) = 2.0
    v(3) = 3.0

    ! calcul du produit scalaire
    scalaire = dot_product(u, v)
    write(*, *) "u.v = ", scalaire

    ! orthogonalisation de schimdt
    if (scalaire /= 0) then
        vp(:) = v(:) - scalaire / normu * u(:)
        ! verification
        write(*, *) "u.vp = ", dot_product(u, vp)
        write(*, *) "vp = ", (vp(i), i = 1, 3)
    end if

end program schmidt

File: src_simple/python/schmidt.py
#!/usr/bin/env python
# -*- coding=utf-8 -*-

""" schmidt  -  13.03.2014 """

from math import sqrt


def schmidt():
    """ Procédé d'orthogonalisation de Gram-Schimdt.
        Soit u et v deux vecteurs. On cherche le vecteur vp le plus
        proche de v orthogonal à u. """

    # vecteur u
    u = [1., 1, 1.]
    normu = sum([ui**2 for ui in u])

    # vecteur v
    v = [1., 2., 3.]

    # calcul du produit scalaire
    scalaire = sum([ui * vi for ui, vi in zip(u, v)])
    print("u.v = {0}".format(scalaire))

    # orthogonalisation de schmidt
    if scalaire != 0.0:
        vp = [vi - scalaire / normu * ui for ui, vi in zip(u, v)]

        # verification
        print("u.vp = {0}".format(sum([ui * vpi for ui, vpi in zip(u, vp)])))
        print("vp = {0}".format(vp))

if __name__ == "__main__":
    schmidt()
